node_modules/oae-core/manageaccess/js/manageaccess.js

```diff
@@ -172,12 +307,38 @@ define(['jquery', 'oae.core', 'jquery.autosuggest'], function($, oae) {
          * @param  {Object[]}    autoSuggestMembers    Trimmed member object as used by the members feed containing all properties necessary to render a list item in the infinite scroll
          */
         var addNewMembers = function(autoSuggestMembers) {
-            $.each(autoSuggestMembers, function(i, newMember) {
+            var membersPartition = _.partition(autoSuggestMembers, function(member) {
+                return (member.profile.resourceType === 'email');
+            });
+
+            var newInvitations = _.first(membersPartition);
+            var newMembers = _.last(membersPartition);
+
+            // Keep track of all members updates, as well as all members we've searched for by
+            // email in order to create validated share targets
+            _.each(newMembers, function(newMember) {
+                if (newMember.email) {
+                    userIdValidations[newMember.profile.id] = newMember.email;
+                }
+
                 membersUpdates[newMember.profile.id] = newMember.role;
                 pendingMembers[newMember.profile.id] = true;
             });
 
-            infinityScroll.prependItems({'results': autoSuggestMembers});
+            // Keep track of all invitations updates
+            _.each(newInvitations, function(newInvitation) {
+                invitationsUpdates[newInvitation.profile.id] = newInvitation.role;
+                pendingInvitations[newInvitation.profile.id] = true;
+            });
+
+            // Update the members infinite scroll and the invitations list
+            if (!_.isEmpty(newInvitations)) {
+                updateInvitations();
```

Comment:
It's a bit more difficult than that because you can wind up with duplicates if the user adds an email that already exists in the list, effectively changing its role. At the moment, it somewhat prepends as it first shows all the "added/updated invitations" at the top of the list, and then appends all the unmodified invitations from the server. However, it doesn't prepend with respect to going back to "Add More" multiple times, the most recent changes will appear at the bottom of all the changes that have been made in that session. I can just reverse that list when updating the listing.